### PR TITLE
[auth] OAuth2 authentication plugin

### PR DESCRIPTION
**Pull request by elpaso**
# Resurrected OAuth2 PR from https://github.com/qgis/QGIS/pull/4437

**OAuth2 authentication method plugin** for QGIS authentication system:
- Supports authentication code, implicit and resource owner grant flows
- Currently defaults to interacting with user's default external browser, though an embedded one could be crafted using `QWebEngineView` ([see caveats on embedded web clients](https://github.com/pipacs/o2/pull/81) for OAuth2 within a native application)
- Allows for preconfigured connections, optionally read from default locations
- Offers caching of access token beyond QGIS restarts (and with linking to `libqt5keychain` library, can securely do so via user's Wallet/Keychain/Password Manager).
- Save/load complicated OAuth2 configurations
- Has documentation written

Work produced by [Boundless Spatial](https://github.com/boundlessgeo) and sponsored by [Monsanto Company](https://github.com/MonsantoCo).

## What's new?

- updated, rebased and tested on current master
- O2 library sources are now in `external`

## Windows test build

A windows build for testing is available at http://www.itopen.it/bulk/mxe-release-2018-06-25-10-10-20.zip (no installation required: just unzip and run qgis, no python enabled)

### CODE REVIEW

File: src/auth/oauth2/qgsauthoauth2edit.cpp [hunk 1]
```diff
@@ -0,0 +1,925 @@
...
+#include "qgsapplication.h"
+#include "qgsauthguiutils.h"
+#include "qgsauthmanager.h"
+#include "qgsauthconfigedit.h"
+#include "qgslogger.h"
+
+
+QgsAuthOAuth2Edit::QgsAuthOAuth2Edit( QWidget *parent )
+  : QgsAuthMethodEdit( parent )
+  , mOAuthConfigCustom( nullptr )
```
Comment (nyalldawson) [inline src/auth/oauth2/qgsauthoauth2edit.cpp]: Move initializations to header

File: src/auth/oauth2/qgsauthoauth2edit.cpp [hunk 2]
```diff
@@ -0,0 +1,925 @@
...
+void QgsAuthOAuth2Edit::initConfigObjs()
+{
+  mOAuthConfigCustom = new QgsAuthOAuth2Config( this );
+  mOAuthConfigCustom->setConfigType( QgsAuthOAuth2Config::Custom );
+  mOAuthConfigCustom->setToDefaults();
+}
+
+void QgsAuthOAuth2Edit::deleteConfigObjs()
+{
+  delete mOAuthConfigCustom;
```
Comment (nyalldawson) [inline src/auth/oauth2/qgsauthoauth2edit.cpp]: Candidate for unique_ptr

File: src/auth/oauth2/qgsauthoauth2edit.cpp [hunk 3]
```diff
@@ -0,0 +1,925 @@
...
+  leClientSecret->setVisible( !implicit );
+  if ( implicit ) leClientSecret->setText( QString() );
+
+  leClientId->setPlaceholderText( resowner ? tr( "Optional" ) : tr( "Required" ) );
+  leClientSecret->setPlaceholderText( resowner ? tr( "Optional" ) : tr( "Required" ) );
+
+
+  lblUsername->setVisible( resowner );
+  leUsername->setVisible( resowner );
+  if ( !resowner ) leUsername->setText( QString() );
```
Comment (nyalldawson) [inline src/auth/oauth2/qgsauthoauth2edit.cpp]: split to multiline (+elsewhere)

File: src/auth/oauth2/qgsauthoauth2edit.cpp (hunk 2 above)
Comment (elpaso) [reply src/auth/oauth2/qgsauthoauth2edit.cpp]: @nyalldawson thanks for the review! I'm looking into this candidate for unique_ptr, it's a QObject so I think we could just leave memory management to Qt (it's parented) and completely drop the dtor, what do you think?

Comment (elpaso) [reply src/auth/oauth2/qgsauthoauth2edit.cpp]: @nyalldawson nevermind, I've done that in latest commit

File: src/auth/oauth2/qgsauthoauth2config.h [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+
+// TODO: add SimpleCrypt or QgsAuthCrypto for (en|de)crypting client secret key
+
+#include <QObject>
+#include <QVariantMap>
+
+#include "qgis.h"
+
+
+class QgsAuthOAuth2Config : public QObject
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/auth/oauth2/qgsauthoauth2config.h]: Good point, I missed that existing providers aren't in dox either.

I'd say ignore my earlier comment and DON'T include these based on that, but just make sure that there's still enough dox included to make the classes understandable for others.

**Comment by 3nids** [comment]
I am no expert here, but mainly pop up for the submodule discussion, I see Nyall already mentionned it.
I would also be in favor of using submodules or any other approach.

Also, maybe I'm off topic, but in Qt 5.10 (preview since 5.8), there is this new feature:
* Qt Network Auth - Provides support for OAuth1 & OAuth2
https://doc.qt.io/qt-5/qtnetworkauth-index.html

Could this be used?

**Comment by elpaso** [comment]
@3nids I'm completely neutral about the approach we choose: the submodule options as well as the CMake subproject have already been discusses in the "old" PR and I remember you comment: https://github.com/qgis/QGIS/pull/4437#issuecomment-311255553

But someone still have concerns about both approaches (submodule and subproject with download) and it seemed to us that the majority of who expressed his opinion was ok with the code inclusion in "external" (which already is used by other external libraries).

So, for now I'll stick to this.

The Qt5 option was also discussed: https://github.com/qgis/QGIS/pull/4437#issuecomment-393419770 and according to @lshaffer who did most of the research on the features, there are some serious limitations in the Qt implementation compared to o2.

**Comment by 3nids** [comment]
thanks a lot for the clarifications and sorry for the noise!

**Comment by dakcarto** [comment]
@elpaso, thanks for all your hard work on resurrecting this authentication plugin!

**Pull request merged**